### PR TITLE
HPB bankruptcy on settle

### PR DESCRIPTION
**Pull request by grandizzy**
- if there's only a tiny amount of deposit (2) backed by 2 LPs in HPB then the collateralUsed to settle is calculated as 0 due to rounding (`vars.collateralUsed = Maths.wdiv(vars.scaledDeposit, vars.price)`)
- this results in removing deposit (2) and not adding any collateral into bucket, leaving an amount of 2 LPs that is not backed by any asset
- fixed by declaring bucket bankruptcy in case settle leaves bucket with 0 collateral, 0 deposit but LPs > 0
- added check for deposit to remove to be min of available deposit in bucket / calculated deposit for settle in order to prevent any potential underflow
- unit test

### CODE REVIEW

**Review by mattcushman — APPROVED** [review]
lgtm

File: src/libraries/external/Auctions.sol [hunk 1]
```diff
@@ -105,15 +105,18 @@ library Auctions {
...
-        uint256 price;             // [WAD] price of settling bucket
-        uint256 scaledDeposit;     // [WAD] scaled amount of quote tokens in bucket
-        uint256 scale;             // [WAD] scale of settling bucket
-        uint256 unscaledDeposit;   // [WAD] unscaled amount of quote tokens in bucket
+        uint256 collateralUsed;     // [WAD] collateral used to settle debt
+        uint256 debt;               // [WAD] debt to settle
+        uint256 depositToRemove;    // [WAD] deposit used by settle auction
+        uint256 hpbCollateral;      // [WAD] amount of collateral in HPB bucket
+        uint256 hpbUnscaledDeposit; // [WAD] unscaled amount of of quote tokens in HPB bucket before settle
+        uint256 hpbLPs;             // [WAD] amount of LPs in HPB bucket
```
Comment (EdNoepel) [inline src/libraries/external/Auctions.sol:113]: note to self: 3 new local vars lines 111-113

File: src/libraries/external/Auctions.sol [hunk 2]
```diff
@@ -217,44 +220,64 @@ library Auctions {
...
+                borrower.collateral -= vars.collateralUsed;
 
-                Deposits.unscaledRemove(deposits_, vars.index, vars.unscaledDeposit); // remove amount to settle debt from bucket (could be entire deposit or only the settled debt)
-            }
+                Bucket storage hpb = buckets_[vars.index];
+                vars.hpbLPs        = hpb.lps;
+                vars.hpbCollateral = hpb.collateral + vars.collateralUsed;
+
+                // set amount to remove as min of calculated amount and available deposit (to prevent rounding issues)
+                vars.unscaledDeposit    = Maths.min(vars.hpbUnscaledDeposit, vars.unscaledDeposit);
```
Comment (EdNoepel) [inline src/libraries/external/Auctions.sol:264]: nit: alignment of `Maths.min` call

File: src/libraries/external/Auctions.sol [hunk 3]
```diff
@@ -217,44 +220,64 @@ library Auctions {
...
                 }
 
-                borrower.collateral             -= vars.collateralUsed;               // move settled collateral from loan into bucket
-                buckets_[vars.index].collateral += vars.collateralUsed;
+                // remove settled collateral from loan
+                borrower.collateral -= vars.collateralUsed;
 
-                Deposits.unscaledRemove(deposits_, vars.index, vars.unscaledDeposit); // remove amount to settle debt from bucket (could be entire deposit or only the settled debt)
-            }
+                Bucket storage hpb = buckets_[vars.index];
```
Comment (EdNoepel) [inline src/libraries/external/Auctions.sol:259]: Appreciate naming this `hpb` and not `settlementBucket`.  I reviewed and confirmed this is always the HPB; the user cannot specify the index.

**Review by EdNoepel — APPROVED** [review]
LGTM

**Pull request merged**